File: 백준/Silver/16194. 카드 구매하기 2/카드 구매하기 2.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var n int
	fmt.Fscanf(reader, "%d\n", &n)
	var cards = make([]float64, n+1)
	var dp = make([]float64, n+1)
	for i := 1; i <= n; i++ {
		fmt.Fscanf(reader, "%f ", &cards[i])
		dp[i] = 0
	}
	dp[0] = 0
	for i := 1; i < n+1; i++ {
		for j := 1; j < i+1; j++ {
			if dp[i] == 0 {
				dp[i] = dp[i-j] + cards[j]
			} else {
				dp[i] = math.Min(dp[i], dp[i-j]+cards[j])
			}
		}
	}
	fmt.Fprintf(writer, "%0.f\n", dp[n])
}

File: 백준/Silver/16194. 카드 구매하기 2/카드 구매하기 2.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n = int(input())
cards = [0]+list(map(int, input().split()))
dp = [0 for _ in range(n+1)]

for i in range(1, n+1):
    for j in range(1, i+1):
        if dp[i] ==0:
            dp[i] = dp[i-j]+cards[j]
        else :
            dp[i] = min(dp[i], dp[i-j]+cards[j])

print(dp[n])